store/tikv/region_request.go

```diff
@@ -182,12 +187,184 @@ func (s *RegionRequestSender) SetRPCError(err error) {
 	s.rpcError = err
 }
 
-// SendReq sends a request to tikv server.
+// SendReq sends a request to tikv server. If fails to send the request to all replicas,
+// a fake region error may be returned. Caller which receives the error should retry the request.
 func (s *RegionRequestSender) SendReq(bo *Backoffer, req *tikvrpc.Request, regionID RegionVerID, timeout time.Duration) (*tikvrpc.Response, error) {
 	resp, _, err := s.SendReqCtx(bo, req, regionID, timeout, tikvrpc.TiKV)
 	return resp, err
 }
 
+type replica struct {
+	store    *Store
+	peer     *metapb.Peer
+	epoch    uint32
+	attempts int
+}
+
+type replicaSelector struct {
+	regionCache *RegionCache
+	region      *Region
+	// replicas contains all TiKV replicas for now and the leader is at the
+	// head of the slice.
+	replicas []*replica
+	// nextReplicaIdx points to the candidate for the next attempt.
+	nextReplicaIdx int
+}
+
+func newReplicaSelector(regionCache *RegionCache, regionID RegionVerID) (*replicaSelector, error) {
+	cachedRegion := regionCache.GetCachedRegionWithRLock(regionID)
+	if cachedRegion == nil || !cachedRegion.isValid() {
+		return nil, nil
+	}
+	regionStore := cachedRegion.getStore()
+	replicas := make([]*replica, 0, regionStore.accessStoreNum(TiKVOnly))
+	for _, storeIdx := range regionStore.accessIndex[TiKVOnly] {
+		replicas = append(replicas, &replica{
+			store:    regionStore.stores[storeIdx],
+			peer:     cachedRegion.meta.Peers[storeIdx],
+			epoch:    regionStore.storeEpochs[storeIdx],
+			attempts: 0,
+		})
+	}
+	// Move the leader to the first slot.
+	replicas[regionStore.workTiKVIdx], replicas[0] = replicas[0], replicas[regionStore.workTiKVIdx]
+	return &replicaSelector{
+		regionCache,
+		cachedRegion,
+		replicas,
+		0,
+	}, nil
+}
+
+// isExhausted returns true if runs out of all replicas.
+func (s *replicaSelector) isExhausted() bool {
+	return s.nextReplicaIdx >= len(s.replicas)
+}
+
+func (s *replicaSelector) nextReplica() *replica {
+	if s.isExhausted() {
+		return nil
+	}
+	return s.replicas[s.nextReplicaIdx]
+}
+
+const maxReplicaAttempt = 10
+
+// next creates the RPCContext of the current candidate replica.
+// It returns a SendError if runs out of all replicas of the cached region is invalidated.
+func (s *replicaSelector) next(bo *Backoffer) (*RPCContext, error) {
+	for {
+		if !s.region.isValid() {
+			return nil, nil
+		}
+		if s.isExhausted() {
+			s.invalidateRegion()
+			return nil, nil
+		}
+		replica := s.replicas[s.nextReplicaIdx]
+		s.nextReplicaIdx++
+
+		// Limit the max attempts of each replica to prevent endless retry.
+		replica.attempts++
+		if replica.attempts > maxReplicaAttempt {
+			continue
+		}
+
+		storeFailEpoch := atomic.LoadUint32(&replica.store.epoch)
+		if storeFailEpoch != replica.epoch {
+			// TODO(youjiali1995): Is it necessary to invalidate the region?
```

Comment:
Maybe only invalidate the replica in this store? The other replicas may become leader and serve.

Reply:
Yes, but the old logic has worked for a long time, I need to benchmark it.